File: src/main/java/persistence/sql/ddl/column/Column.java
package persistence.sql.ddl.column;

import java.lang.reflect.Field;


public class Column {

    private final ColumnName name;
    private final ColumnType type;
    private final ColumnLength length;
    private final boolean nullable;

    private Column(ColumnName name, ColumnType type, ColumnLength length, boolean nullable) {
        this.name = name;
        this.type = type;
        this.length = length;
        this.nullable = nullable;
    }

    public static Column from(Field field) {
        jakarta.persistence.Column column = field.getAnnotation(jakarta.persistence.Column.class);

        ColumnName name = ColumnName.from(field);
        ColumnType type = ColumnType.findColumnType(field.getType());
        ColumnLength length = ColumnLength.from(field);
        boolean nullable = column != null && column.nullable();

        return new Column(name, type, length, nullable);
    }

    public String getLengthDefinition() {
        if (type != ColumnType.STRING) {
            return "";
        }

        return String.format("(%d)", length.getLength());
    }

    public String getNullableDefinition() {
        if (nullable) {
            return "null";
        }

        return "not null";
    }

    public String getName() {
        return name.getName();
    }

    public String getType() {
        return type.getMysqlColumnType();
    }
}

File: src/main/java/persistence/sql/ddl/column/Column.kt
package persistence.sql.ddl.column

import jakarta.persistence.Transient
import parse.ParseType
import persistence.sql.find
import java.lang.reflect.Field

data class Column (
    private val columnParsers: Map<ParseType, ColumnParser>,
    private val field: Field
) {

    fun createQuery(): String {
        if (isTransientField()) {
            return ""
        }

        return listOf(
            columnParsers.find(ParseType.NAME).parse(field),
            columnParsers.find(ParseType.TYPE).parse(field),
            columnParsers.find(ParseType.NULLABLE).parse(field),
            columnParsers.find(ParseType.ID).parse(field))
        .filter { it.isNotEmpty() }
        .joinToString(" ")
    }

    private fun isTransientField(): Boolean {
        return field.isAnnotationPresent(Transient::class.java)
    }
}
